src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,486 @@
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use async_trait::async_trait;
+use tokio::sync::oneshot::{Receiver, Sender};
+
+use crate::{
+    LocalUriProvider, UCode, UListener, UMessage, UMessageBuilder, UStatus, UTransport, UUri, UUID,
+};
+
+use super::{CallOptions, RpcClient, ServiceInvocationError, UPayload};
+
+fn handle_response_message(response: UMessage) -> Result<Option<UPayload>, ServiceInvocationError> {
+    if let Some(attribs) = response.attributes.as_ref() {
+        match attribs.commstatus.map(|v| v.enum_value_or_default()) {
+            Some(UCode::OK) | None => {
+                // successful invocation
+                response.payload.map_or(Ok(None), |payload| {
+                    Ok(Some(UPayload::new(
+                        payload,
+                        attribs.payload_format.enum_value_or_default(),
+                    )))
+                })
+            }
+            Some(code) => {
+                // try to extract UStatus from response payload
+                let status = response.extract_protobuf().unwrap_or_else(|_e| {
+                    UStatus::fail_with_code(code, "failed to invoke service operation")
+                });
+                Err(ServiceInvocationError::from(status))
+            }
+        }
+    } else {
+        Err(ServiceInvocationError::RpcError(UStatus::fail_with_code(
+            UCode::INTERNAL,
+            "response message does not contain attributes",
+        )))
+    }
+}
+
+struct ResponseListener {
+    // request ID -> sender for response message
+    pending_requests: Mutex<HashMap<UUID, Sender<UMessage>>>,
+}
+
+impl ResponseListener {
+    fn try_add_pending_request(
+        &self,
+        reqid: UUID,
+    ) -> Result<Receiver<UMessage>, ServiceInvocationError> {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            if let Entry::Vacant(entry) = pending_requests.entry(reqid) {
+                let (tx, rx) = tokio::sync::oneshot::channel();
+                entry.insert(tx);
+                Ok(rx)
+            } else {
+                Err(ServiceInvocationError::AlreadyExists(
+                    "RPC request with given ID already pending".to_string(),
+                ))
+            }
+        } else {
+            Err(ServiceInvocationError::RpcError(UStatus::fail_with_code(
+                UCode::INTERNAL,
+                "failed to add response handler",
+            )))
+        }
+    }
+
+    fn handle_response(&self, reqid: &UUID, response_message: UMessage) {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
```

Comment:
Same concept here about avoiding indentation and preferring early return. May just be a style thing? For your consideration:

```rust
let mut pending_requests = self.pending_requests.lock() else {
  // TODO: May be good to explicitly log some sort of error here stating that the Mutex has been poisoned?
  // Reference: https://doc.rust-lang.org/std/sync/type.LockResult.html
  return;
}

let Some(sender) = pending_requests.remove(reqid)  else {
  // we seem to have received a duplicate of the response message,
  // ignoring it
  // TODO: May be good to explicitly log that we already handled the previous duplicate response message?
  return;
}

let _ = sender.send(response_message);
```

Reply:
what level would you log this at, DEBUG?

Reply:
Yeah, that seems reasonable to me.